### DISCUSSION TITLE
How to manage users permissions with Nutgram ?

### DISCUSSION THREAD

**Discussion by ghost** [nutgram/nutgram#297] [Q&A] [answered]
Hi

How can I check the permissions of users who are registered in the bot as admin in Database?

I can do it with the code below. But I need a better solution so that I don't have to define a middleware for each command

```php
use App\Models\User;
use SergiX44\Nutgram\Nutgram;

$bot = new Nutgram($token);

$bot->middleware(function (Nutgram $bot) {
    $user = User::find(1);
    $bot->setData($user, 'user');
});

$bot->onCommand('add_channel', function (Nutgram $bot) {
    // ...
})->middleware(function (Nutgram $bot, $next) {
    $user = $bot->getData('user');
    if (!$user->permissions->can_add_channel) {
        throw new Exception("You don't have permission to add channel");
    }
    $next($bot);
});

$bot->onCommand('add_image', function (Nutgram $bot) {
    // ...
})->middleware(function (Nutgram $bot, $next) {
    $user = $bot->getData('user');
    if (!$user->permissions->can_add_image) {
        throw new Exception("You don't have permission to add image");
    }
    $next($bot);
});

$bot->run();
```

-----------------------------------------------------------------------------

I'm thinking of just creating a middleware called ```CheckUserPermissionMiddleware``` and passing the class name to the relevant commands with some values to that middleware, but I don't know how to do it. What I want is similar to the following code:

```CheckUserPermissionMiddleware :```
```php
namespace App\Middlewares;

use SergiX44\Nutgram\Nutgram;

class CheckUserPermissionMiddleware
{
    // pass permission key or index to middleware ( can_add_channel or can_add_image )
    public function __invoke(Nutgram $tg, $next, $permissionKey) 
    {
        $user = $tg->getData('user');
        if (! $user->permissions->{$permissionKey}) {
            throw new Exception("You don't have permission to do this!");
        }
        $next($tg);
    }
}
```

```Main :```
```php
use App\Models\User;
use App\Middlewares\CheckUserPermissionMiddleware;
use SergiX44\Nutgram\Nutgram;

$bot = new Nutgram($token);

$bot->middleware(function (Nutgram $bot) {
    $user = User::find(1);
    $bot->setData($user, 'user');
});

$bot->onCommand('add_channel', function (Nutgram $bot) {
    // ...
})->middlewareWithAValue(CheckUserPermissionMiddleware::class, 'can_add_channel');

$bot->onCommand('add_image', function (Nutgram $bot) {
    // ...
})->middlewareWithAValue(CheckUserPermissionMiddleware::class, 'can_add_image');

$bot->run();
```

I hope you understand what I mean.

**How can I implement this correctly in Nutgram?**

**Comment by juankstylez** [comment]
Good idea ! I would like to know it too! I will be attentive to the answer!


A question apart from everything: how are you doing to log in to the users? I'm looking for a way to use the same auth system as Laravel! but I can't find the way!

How are you registering and logging users?

**Reply by ghost** [reply]
I am new to Laravel and its authentication system. (Actually, I don't know anything about Laravel's authentication system and I don't use Laravel to develop Telegram bots)

I only use Nutgram and Eloquent as a package in my pure PHP project.

Regarding your last question, my bot does not have the ability to login users. I just define some admins with specific permissions in my bot.

It's just difficult for me to check admins permissions to different sections of the bot in Nutgram (of course, it's possible to implement it, but I don't think the solution I proposed in this conversation is interesting and the codes are a bit more and out of cleanliness)

I hope the Nutgram developers will design a useful system for this or provide a suggestion to better manage admins/users permissions in the bot.

**Reply by juankstylez** [reply]
I understand, in my case for now I have not been able to implement laravel authentication in my bot where I am learning since laravel works with an authentication token.

For now, what I did was bring the user's data associating them with the telegram ID xxxxxxxxxx, create a tgid field in the database and when the user starts the bot, it asks for their name, email and this is automatically associated with their tg ID, for if you ever need it ;) from then on you can use eloquent to bind to permissions and more.

regards

**Comment by Lukasss93** [comment, marked as answer]
But... wait!

Just do this:

```php

// CheckUserPermissionMiddleware.php
class CheckPermissions
{
    protected string $permission;

    public function __construct(string $permission)
    {
        $this->permission = $permission;
    }

    public function __invoke(Nutgram $bot, $next): void
    {
        //check if user has permssion using $this->permission

        $next($bot);
    }
}

// routes:
use App\Models\User;
use App\Middlewares\CheckUserPermissionMiddleware;
use SergiX44\Nutgram\Nutgram;

$bot = new Nutgram($token);

$bot->middleware(function (Nutgram $bot) {
    $user = User::find(1);
    $bot->setData($user, 'user');
});

$bot->onCommand('add_channel', function (Nutgram $bot) {
    // ...
})->middleware(new CheckUserPermissionMiddleware('can_add_channel'));

$bot->onCommand('add_image', function (Nutgram $bot) {
    // ...
})->middleware(new CheckUserPermissionMiddleware('can_add_image'));

$bot->run();






```

**Reply by ghost** [reply]
It looks great!
Thanks bro :)
